### PR TITLE
feat(ltx2): add the lip-dub pipeline for re-voicing a reference clip

### LINKED ISSUE #595: feat(ltx2): lip-dub pipeline (reference-video re-voicing)
Upstream's `LipDubPipeline` re-voices a reference video (rephrase while keeping speaker identity) using a dedicated lip-dub IC-LoRA. mold has the ingredients — source-video conditioning, audio conditioning (`a2vid`), IC-LoRA mode — but no lip-dub pipeline wiring, adapter manifest, or reference-video + new-audio request shape.

**Scope**
- `Ltx2PipelineMode::LipDub` taking `--reference-video` + audio (file or generated from prompt), with upstream's frame-count snapping rules.
- Manifest/catalog entry for the lip-dub adapter; CLI + web surface.

Depends on the IC-LoRA control-adapter groundwork (manifest/catalog patterns for adapters).

### PR DESCRIPTION
**Pull request by jamesbrink**
Closes #595.

Upstream's `LipDubPipeline` takes a clip of someone speaking and a new line of dialogue and re-renders the speaker saying it, keeping their voice and their face. mold already had every ingredient — source-video conditioning, an audio VAE, IC-LoRA — so what was missing was the pipeline itself. #672 registered the gated `Lightricks/LTX-2.3-22b-IC-LoRA-DubIt` adapter; this gives it something to run.

## Three things that are not like the other pipelines

**The reference clip owns the timeline.** Frames and frame rate are read off the source stream and rounded *down* onto the `8k+1` grid, replacing whatever the request asked for (upstream `lipdub.py:46-49`, `:190-192`). A dub rendered on a different timeline is out of sync, and an out-of-sync dub looks perfectly fine in a thumbnail — so the substitution is **reported**, never silent: an SSE `Info` event and the response header on the server, a progress line on the CLI.

It is resolved in `prepare_generation`, *before* validation and before the scheduler prices the job. That placement is the point: a 20-second reference admitted against a 97-frame VRAM grant would be five times the size it was measured at. `materialize_request` derives the same numbers through the same `mold_core::validation::resolve_lip_dub_timing`, so a forced-local run — which never sees the server — lands on the identical timeline rather than a second implementation of the same rule.

**The reference voice is appended, not mixed.** `StageAudioConditioning` is the audio mirror of the existing `StageVideoConditioning`: the clip's own speech is encoded through the audio VAE and appended to the denoised sequence as tokens the transformer may attend to but never denoises. Everything it needs already existed — `AudioPatchifier` is bit-identical to upstream's einops, `get_patch_grid_bounds` already emits seconds with the causal offset, and the transformer's `audio_self_attention_mask` slot was already there and hardcoded `None`. This is what turns it on.

The load-bearing detail is the RoPE shift. Reference tokens carry the same seconds-valued positions the generated audio would, so without a shift the model sees two soundtracks stacked on one timeline. Upstream subtracts the reference's own duration plus exactly one audio latent frame — `positions -= last_end + 0.04` (`lipdub.py:314-315`) — leaving the last reference patch ending at exactly `-0.04 s`, immediately before the generated audio's `0.0`. Getting this wrong produces output that is plausible *and* desynchronized, which is the worst failure mode available, so it is pinned by an exact assertion at `T ∈ {1, 3, 126}` rather than a range check.

**Stage 2 is a picture-only refinement.** Generic IC-LoRA builds its second stage with `loras=()` and conditions it on images only (`ic_lora.py:99-103`, `:279-288`). Lip dub does the opposite on four counts, each a named test:

| | generic IC-LoRA | lip dub | upstream |
| --- | --- | --- | --- |
| adapter in stage 2 | dropped | **kept** | `lipdub.py:96-106` (one `DiffusionStage`, both passes) |
| reference video in stage 2 | dropped | **kept** | `lipdub.py:236`, `:265` |
| stage-2 audio | re-noised and denoised | **frozen** on the stage-1 latent | `lipdub.py:283-289` |
| stage-2 audio reference | — | rebuilt from the **generated** audio | `lipdub.py:267` |
| exported track | stage 2 | **stage 1** | `lipdub.py:293` |

Re-denoising a finished dub can only move it away from the mouth already rendered against it.

## Surface

`--pipeline lip-dub`, or just `--ic-lora-control lipdub`, which selects it. Every surface derives that routing from one rule — `mold_core::ltx2_control::pipeline_for_control_id`, mirrored for the browser in `studio/lib/ltx2Control.ts` and pinned to the Rust registry by a parity test. Asking for the lip-dub adapter on `ic-lora` is refused rather than honoured: it would load the right weights and run the wrong graph.

Width and height must be multiples of 64 — the pipeline is always two-stage, so stage 1 halves both axes (upstream `assert_resolution(..., is_two_stage=True)`) — and the reference must carry an audio track, because its speech is what the dub imitates. Every such precondition lives in `resolve_lip_dub_timing`, which takes a `LipDubReference { frames, fps, has_audio }`: a reference that cannot drive a dub is refused at the request boundary rather than after a 22B checkpoint has loaded, and adding a precondition is a compile error at both the server and forced-local call sites so they cannot drift.

Responses gain `x-mold-request-warning`, carrying every advisory about an accepted request. `x-mold-dimension-warning` keeps its documented meaning and still carries dimension adjustments only — a timing substitution routed through it is one a client could reasonably filter out, which would defeat the point of reporting it at all.

Web, desktop, and iPhone Advanced → Video gain the mode, the routing, and a hint saying where the timing comes from.

## Verification

Real render on an RTX 4090 (24 GB), `ltx-2.3-22b-distilled:fp8`, against a 704×448 / 24 fps / 25-frame reference carrying a non-silent AAC track:

```
reference.mp4  h264 704x448 24/1 25 frames · aac 48000 Hz 2ch   mean -22.6 dB  max -6.5 dB
dub.mp4        h264 704x448 24/1 25 frames · aac 48000 Hz 2ch   mean -13.5 dB  max -0.8 dB
```

Width, height, frame count, and frame rate match the reference exactly. The audio is present, non-silent, and *different* from the reference's, so it is generated rather than passed through.

Not a placeholder: the runtime logged `render_native_video using real path pipeline=LipDub`, and a mid-clip frame is a photographic close-up of the speaker — same subject, framing, and background as the reference, with the mouth re-timed — not the synthetic gradient.

The mechanisms are visible in the render log too. Stage 1 ran `[1, 52, 128]` audio tokens (26 generated + 26 appended reference) and `[1, 2464, 128]` video tokens (1232 + 1232 reference, `reference_downscale_factor = 1` read from the adapter's own metadata). Stage 2 ran `[1, 9856, 128]` video tokens — 4928 + 4928 — so the reference survived into the refinement, and its `step_audio_latents` were byte-identical across all three steps (`mean=0.104014 abs_mean=0.982685 rms=1.275666`), so the audio really was frozen.

## Gates

`cargo fmt` · `cargo clippy --workspace --all-targets -D warnings` · `cargo test --workspace` · feature-combo check · studio/web/desktop fmt, typecheck, tests, and builds · `check:architecture` · `check:dead-code` · iOS asset scripts · `website` fmt/verify/build.

`cargo test --workspace` is **clean** — 5305 passing, zero failures — now that #682 has landed and fixed the artifact-identity cluster.

The one caveat left is environmental: `desktop/src/router.test.ts` intermittently times out at 5000 ms on a lazy route import when this machine is at a load average of ~119 (several agents' suites in parallel). It reproduces on unmodified `origin/main` and passes in isolation, so it is load, not code.

No new blocking waits: the only `await` this branch adds is a config `RwLock` read. Nothing here subscribes to the download channel, so there is no path that can stall a test the way the pre-`control_missing_files` completeness check could.

## Out of scope

- **The TUI** keeps its existing advanced-video gap (tracked in #600); it exposes no pipeline selector.
- **Discord** deliberately still lists only the four input-free pipelines — lip dub needs a source video the slash command does not collect.
- **Chained sequences** do not offer lip dub. A chain clip has no reference video of its own, and the per-clip timing would fight the reference's.
- **Spatial and temporal upscale** are rejected (422) with lip dub rather than silently ignored — both would change the output shape out from under the clip the dub has to line up with, and upstream's pipeline composes with neither.
- **Dub quality** was not evaluated. The reference used for verification is a generated clip whose audio is ambient rather than clean speech, so this proves the pipeline runs end to end on real weights and produces real, correctly-shaped, non-silent output — not that a particular line reads well.

https://claude.ai/code/session_01HSPYvNCPvWLRTBYbit2CHv

### CODE REVIEW

File: crates/mold-server/src/routes.rs [hunk 1]
```diff
@@ -905,6 +915,57 @@ async fn plan_builtin_ltx2_control(
...
+        // No reference at all: `validate_generate_request` says so far better
+        // than a probe failure would.
+        return Ok(Vec::new());
+    };
+    let probe = probe.map_err(|e| {
+        ApiError::validation(format!("could not read the lip-dub reference video: {e:#}"))
+    })?;
+    let frames = probe.frames.ok_or_else(|| {
+        ApiError::validation("the lip-dub reference video reports no frame count")
+    })?;
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply crates/mold-server/src/routes.rs:984]: Fixed — and thank you, this was a real hole.

Rather than add the check at this call site, I moved it into the shared resolver so the two paths cannot diverge. `resolve_lip_dub_timing` now takes a `LipDubReference { frames, fps, has_audio }` struct instead of positional arguments, and rejects a silent reference itself:

```rust
if !has_audio {
    return Err(
        "lip-dub reference video has no audio track; the pipeline re-voices existing \
         speech, so the reference must contain some"
            .to_string(),
    );
}
```

Two reasons for the struct rather than a fourth positional argument: `resolve_lip_dub_timing(frames, fps, has_audio, ...)` is a readability trap at the call site, and — the real reason — adding a field is now a compile error at *every* caller, which is the property that keeps the server and forced-local paths validating the same thing. That is the same argument the PR already makes for deriving the timing in one place; you are right that it applies to the audio track too.

Because it lives in the resolver, this now fires in `prepare_generation` before validation, scheduling, and the VRAM grant — so a silent reference is a 422, not a job that dies after a 22B checkpoint has loaded. `lip_dub_timing_rejects_unusable_references` covers it alongside the fps and minimum-length cases.

The defence-in-depth error in `load_lip_dub_reference_audio_latents` stays as the last line for any future caller that reaches the runtime without going through the resolver.

File: crates/mold-inference/src/ltx2/pipeline.rs [hunk 2]
```diff
@@ -403,6 +404,36 @@ impl Ltx2Engine {
...
+            PipelineKind::LipDub => {
+                let reference = conditioning.video_path.as_deref().context(
+                    "the LTX-2 lip-dub pipeline requires a reference video (source_video)",
+                )?;
+                let probe = media::probe_video(Path::new(reference)).with_context(|| {
+                    format!("failed to read the lip-dub reference video '{reference}'")
+                })?;
+                let frames = probe.frames.with_context(|| {
+                    format!("lip-dub reference video '{reference}' reports no frame count")
+                })?;
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply crates/mold-inference/src/ltx2/pipeline.rs:464]: Fixed, same change as the server-side thread (`routes.rs:953`) — deliberately one change rather than two.

The forced-local path now builds the same `LipDubReference { frames, fps: probe.fps, has_audio: probe.has_audio }` and hands it to `resolve_lip_dub_timing`, which owns the rejection. Duplicating the check at both call sites was the obvious fix and the wrong one: it is exactly the shape that lets the two paths drift, which is what this PR was already trying to avoid for the frames/fps derivation. Now the precondition exists once and adding another is a compile error at both callers.

File: crates/mold-server/src/routes.rs [hunk 3]
```diff
@@ -1418,13 +1479,14 @@ async fn generate(
                     })?,
                 );
             }
-            if let Some(warning) = dim_warning {
-                match HeaderValue::from_str(&warning.replace('\n', " ")) {
+            if !warnings.is_empty() {
+                let joined = warnings.join("; ").replace('\n', " ");
+                match HeaderValue::from_str(&joined) {
                     Ok(val) => {
                         headers.insert("x-mold-dimension-warning", val);
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply crates/mold-server/src/routes.rs]: Agreed, and fixed by adding the new header rather than redefining the old one.

This mattered more than a naming nit: the whole argument for the timing substitution is that it must be *reported and never silent*, and shipping it through a header a client may reasonably filter as "dimension stuff" defeats that.

`prepare_generation` now returns a typed `RequestWarnings { dimension, other }` instead of a flat `Vec<String>`, so the two kinds stay distinguishable all the way to the response:

- `x-mold-dimension-warning` — dimension adjustments only, exactly as documented. Unchanged for every existing client.
- `x-mold-request-warning` — every advisory about an accepted request, `;`-separated, request-specific ones first.

Both are in the CORS `expose_headers` list, and the SSE paths emit one `info` event per advisory as before.

Docs updated in `website/api/index.md` (which previously documented only the dimension header) and `desktop/docs/server-api.md`. `request_warnings_keep_dimension_advisories_in_their_own_channel` pins the split, including the case that motivated it — a timing-only warning must leave `dimension` empty rather than publishing itself as a dimension adjustment.

File: crates/mold-inference/src/ltx2/media.rs [hunk 4]
```diff
@@ -579,7 +585,21 @@ pub fn write_contact_sheet(input_video: &Path, output_png: &Path) -> Result<()>
...
+/// [`probe_video`] for a clip that has not been written to disk yet — the
+/// server needs a lip-dub reference's frame count and rate before it stages
+/// the inline `source_video` body anywhere.
+///
+/// Copies the body once, because MP4 parsing needs owned random access. That
+/// is bounded by `MAX_INLINE_SOURCE_VIDEO_BYTES` (64 MiB) and happens once per
+/// request; a server-side `source_video_path` avoids it entirely.
+pub fn probe_video_bytes(bytes: &[u8]) -> Result<ProbeMetadata> {
+    probe_reader(read_mp4_bytes(bytes.to_vec())?)
+}
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply crates/mold-inference/src/ltx2/media.rs:600]: Fixed, and you were right about the comment being the worse half of this.

I checked before deciding: `find_video_track`, `audio_metadata`, and `probe_reader` all took the concrete `&Mp4Reader<Cursor<Vec<u8>>>`, so the copy really was a signature constraint and not an inherent requirement of MP4 parsing. My comment asserted the opposite, which is worse than no comment — it would have told the next reader the allocation was unavoidable and closed the question.

Since it was avoidable, I removed the copy rather than documenting it. The three helpers are now generic over `R: Read + Seek`, and `probe_video_bytes` borrows:

```rust
fn read_mp4_slice(bytes: &[u8]) -> Result<Mp4Reader<Cursor<&[u8]>>> {
    Mp4Reader::read_header(Cursor::new(bytes), bytes.len() as u64)
        .context("failed to parse MP4 container from the request body")
}
```

`read_mp4` (the path-based one) is untouched — `Cursor<Vec<u8>>` still satisfies the new bounds. The comment now says only what is true: it borrows because an inline `source_video` can be 64 MiB and the header parse only seeks around it.

File: crates/mold-inference/src/ltx2/runtime.rs [hunk 5]
```diff
@@ -2067,6 +2297,71 @@ fn append_conditioning_attention_mask(
...
+    let mut values = Vec::with_capacity(conditioning.appended_token_count()?);
+    for condition in &conditioning.appended {
+        values.extend(std::iter::repeat_n(
+            (1.0 - condition.strength) as f32,
+            condition.tokens.dim(1)?,
+        ));
+    }
+    let appended = Tensor::from_vec(values.clone(), (1, values.len()), device)?
+        .broadcast_as((audio_shape.batch, values.len()))?
+        .contiguous()?;
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply crates/mold-inference/src/ltx2/runtime.rs]: Fixed in `build_audio_conditioning_denoise_mask`:

```rust
let appended_tokens = values.len();
let appended = Tensor::from_vec(values, (1, appended_tokens), device)?
    .broadcast_as((audio_shape.batch, appended_tokens))?
```

Note for anyone reading this later: the identical `values.clone()` two functions up in `build_video_conditioning_denoise_mask` is pre-existing, and I left it alone deliberately rather than sweep it in — it is unrelated to lip dub and would put an untested hunk in a feature PR. Worth a follow-up if anyone is tidying that module.
